Updated HBase BulkMutation to use new Batcher API through GCJ client

rahulKQL opened the pull request:
This change will enable `BigtableBufferMutator` to use GCJ's `Batcher<>` class through IBulkMutation when user set `useGCJClient`/`google.bigtable.use.gcj.client` flag.

### Summary of changes it contians: 
 - Had to update `IBigtableDataClient#createBulkMutationBatcher(tableID)` to accept tableID which is need for `Batcher` creation.
 - Updated `IBulkMutation` to adapt against `RowMutationEntry`.
 - Now `BulkMutationGCJClient` uses `Batcher<>` to perform batching.
 - Updated `HBaseRequestAdapter` to adapt against `RowMutationEntry` & also incorporate `RowMutationEntry#createUnsafe()`. 
 - Fixed Junit for the Batching related changes and added test case in IT test class `TestBufferedMutator`.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/core/IBigtableDataClient.java:
```diff
@@ -67,7 +67,7 @@
   ApiFuture<Row> readModifyWriteRowAsync(ReadModifyWriteRow readModifyWriteRow);
 
   /** Creates {@link IBulkMutation} batcher. */
-  IBulkMutation createBulkMutationBatcher();
```
Please mark this as InternalOnly

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/core/IBulkMutation.java:
```diff
@@ -16,28 +16,34 @@
+ * Interface to support batching multiple {@link RowMutationEntry} request into a single grpc
+ * request.
  */
 public interface IBulkMutation {
```
Please mark this class as InternalOnly

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGCJClient.java:
```diff
@@ -69,8 +69,8 @@ public Row readModifyWriteRow(ReadModifyWriteRow readModifyWriteRow) {
   }
 
   @Override
```
Please mark this entire class as InternalOnly

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationGCJClient.java:
```diff
@@ -16,35 +16,36 @@
- * Google-Cloud-Java's {@link BulkMutationBatcher} api.
+ * Google-Cloud-Java's {@link Batcher} api.
  */
 public class BulkMutationGCJClient implements IBulkMutation {
```
InternalOnly

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationGCJClient.java:
```diff
@@ -16,35 +16,36 @@
     final ApiFuture<Void> response = bulkMutateBatcher.add(rowMutation);
+    // TODO(rahulkql): Still using operationAccount to give user isFlush() functionality. which
+    //  is not present in Batcher.
     operationAccountant.registerOperation(ApiFutureUtil.adapt(response));
```
Please avoid using operationAccountant. It seems like this functionality is only used for the jvm shutdown hook to make sure there are no outstanding batches. We have this functionality in gcj already. Would it be possible move the shutdown hook to BulkMutationWrapper?

rahulKQL replied:
By shutdown hook did you meant [this](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutationBatcher.java#L72-L84)? 

Also, I believe we don't need to wait till all the elements are finished as `Batcher#close` is going to flush it out. Please let me know If I am missing something here.

To provide `BulkMutationGCJClient#isFlush()` functionality I have used an AtomicLong#compare(0, 0), which I think should work. Please let me know if there could be a better approach?

igorbernstein2 replied:
If I'm not mistaken, the only place that needs isFlushed() is:

https://github.com/googleapis/java-bigtable-hbase/blob/c4e970d347746f11aff082d3e87787b7e74eb59f/bigtable-client-core-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableConnection.java#L59-L77

And since we already check for unclosed batchers in gcj, I'm proposing removing the operation accountant from the gcj wrapper.

Please let me know if I'm missing something

rahulKQL replied:
Yes, this was that is the only place where it has been used. I did remove it from `BulkMutationGCJClient` and replaced it with AtomicLong check. Please let me know if that looks good to you.

rahulKQL commented on bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestBulkMutationGCJClient.java:
```diff
@@ -17,114 +17,125 @@
+              }
+            })
+        .when(batcher)
+        .flush();
```
Should we make actuals calls to verify this OR this mock test would suffice?

igorbernstein2 replied:
I think it should suffice

kolea2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/core/IBigtableDataClient.java:
```diff
@@ -69,7 +69,7 @@
 
   /** Creates {@link IBulkMutation} batcher. */
-  IBulkMutation createBulkMutationBatcher();
+  IBulkMutation createBulkMutationBatcher(String tableId);
```
we should be careful here as this is breaking. Can we deprecate existing functionality and use this and an overloaded method for now?

rahulKQL replied:
To provide backward compatibility for existing flow, we might need to add an author `IBulkMutation` class (let's say `BulkMutationGCJClientV2`). As the existing wrapper expects `BulkMutationBatcher` and in this change, we are migrating it to `Batcher<>`.
https://github.com/googleapis/java-bigtable-hbase/blob/276834cde845c9a7efee81fb682f5185d5d979b9/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationGCJClient.java#L42-L45

I think `IBigtableDataClient` suppose to be an internal interface. If anyone is using that then it would be great for them to migrate to GCJ veneer library instead of using bigtable-core client.

If you feel we should not break this internal API then we can add flow with existing methods.

igorbernstein2 replied:
Agreed. These wrappers only exist to abstract the api for the hbase client. All of them should be marked as `@InternalOnly`

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/core/IBulkMutation.java:
```diff
@@ -17,28 +17,29 @@
+  ApiFuture<Void> add(RowMutationEntry rowMutation);
 
-  /** Sends any outstanding {@link RowMutation}s, present in the current batch. */
+  /** Sends asynchronously any outstanding RowMutationEntry, present in the current batch. */
```
It synchronously sends, but doesn't wait for the responses

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutationGCJClient.java:
```diff
@@ -17,54 +17,48 @@
   public void sendUnsent() {
-    LOG.info("This operation will be implemented once the underlying API has this feature.");
+    try {
+      bulkMutateBatcher.flush();
```
shouldnt this be non-blocking?

rahulKQL replied:
Do you think I should execute it in the separate thread something like:
```java
  /** {@inheritDoc} */
  @Override
  public void sendUnsent() {
    new Thread() {
      @Override
      public void run() {
        try {

          bulkMutateBatcher.flush();
        } catch (InterruptedException ex) {
          throw new RuntimeException("Could not complete RPC for current Batch", ex);
        }
      }
    }.start();
  }

```

igorbernstein2 replied:
I just meant you should use sendOutstanding() instead of flush. Unless I'm mistaken this method is used for async operations in hbase that return a future and we just needed something start the request

igorbernstein2 replied:
ping

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/HBaseRequestAdapter.java:
```diff
@@ -145,10 +146,10 @@ public void adapt(Delete delete, MutationApi<?> mutationApi) {
    *
    * @param delete a {@link Delete} object.
-   * @return a {@link RowMutation} object.
+   * @return a {@link RowMutationEntry} object.
```
Please mark this class as InternalOnly

rahulKQL commented:
Thanks for detailed instruction, attaching these details for information purpose.

After running the stress test, these are my observations

| No. of rows/thread | Bigtable-core | GCJ-Client |
| ------------------ | ------------- | ---------- |
|     100_000        |      52_450   |    53_787  |
|    10*100_000      |     196_777   |    91_832  |
|    25*100_000      |     432_179   |   190_677  |
|    40*100_000      |     677_109   |   288_978  |
|    50*100_000      |     858_070   |   363_713  | 

<img width="609" alt="Final_batching_chart" src="https://user-images.githubusercontent.com/42969463/65273364-9f628500-db3e-11e9-828d-58280025cfba.png">

sduskis commented:
@rahulKQL, just to clarify, are you saying that GCJ is 2X more performant than Bigtable-core?

rahulKQL commented:
It certainly seems 😃.

Have just added the test case I ran for the observation. @sduskis Would you please take a look?

Also, adding the script log for the case of 2500K & 4000K(forgot to copy the logs for 50 as I ran that first).
[row_40_100_000.log](https://github.com/googleapis/java-bigtable-hbase/files/3633014/row_40_100_000.log)
[row_25_100_000.log](https://github.com/googleapis/java-bigtable-hbase/files/3633015/row_25_100_000.log)

igorbernstein2 commented:
Thats good news. Please update the sendOutstanding issue and we can merge this

rahulKQL commented:
@igorbernstein2 Thanks for the clarification, I have removed the stress test from this PR and added a TODO.

Please have a look and let me know if I can merge this.

igorbernstein2 commented:
LGTM, but we need to wait for @kolea2 to approve

rahulKQL commented:
@kolea2 If you get time, please take a look.

igorbernstein2 commented:
Hi Rahul,

We discussed this last week and decided that we need exercise a bit more caution before merging this. The current plan is:
1. Annotate all classes & interfaces in this repo with appropriate gax visibility annotations (`@InternalOnly`, `@InternalExtensionOnly, etc)
2. create a separate branch `v2.x`
3. merge this PR into that branch
4. finish the migration to gcj
5. release 2.0-alpha with a runtime flag to switch between implementations
6. release 2.0-beta with gcj only

igorbernstein2 commented:
I just opened PRs for the annotations: #2245, #2246, #2247. Once we merge those, we will create a branch where we can merge this PR

igorbernstein2 commented:
can you re-target  this PR to https://github.com/googleapis/java-bigtable-hbase/tree/bigtable-client-2.x?

rahulKQL commented:
@igorbernstein2 Thanks for the notifying me, Have rebased and updated the target branch to `bigtable-client-2.x`.
